frame/staking/src/pallet/mod.rs

```diff
@@ -932,29 +932,50 @@ pub mod pallet {
 		/// the funds out of management ready for transfer.
 		///
 		/// No more than a limited number of unlocking chunks (see `MaxUnlockingChunks`)
-		/// can co-exists at the same time. In that case, [`Call::withdraw_unbonded`] need
-		/// to be called first to remove some of the chunks (if possible).
+		/// can co-exists at the same time. If there are no unlocking chunks slots available
+		/// [`Call::withdraw_unbonded`] is called to remove some of the chunks (if possible).
 		///
 		/// If a user encounters the `InsufficientBond` error when calling this extrinsic,
 		/// they should call `chill` first in order to free up their bonded funds.
 		///
 		/// Emits `Unbonded`.
 		///
 		/// See also [`Call::withdraw_unbonded`].
-		#[pallet::weight(T::WeightInfo::unbond())]
+		#[pallet::weight(
+            T::WeightInfo::withdraw_unbonded_kill(SPECULATIVE_NUM_SPANS).saturating_add(T::WeightInfo::unbond()))
+        ]
 		pub fn unbond(
 			origin: OriginFor<T>,
 			#[pallet::compact] value: BalanceOf<T>,
-		) -> DispatchResult {
+		) -> DispatchResultWithPostInfo {
 			let controller = ensure_signed(origin)?;
+			let unlocking = Self::ledger(&controller)
+				.map(|l| l.unlocking.len())
+				.ok_or(Error::<T>::NotController)?;
+
+			// ensure that there's chunk slots available by requesting the staking interface to
+			// withdraw chunks older than `BondingDuration`, if there are no more unlocking chunks
+			// slots available.
+			let maybe_withdraw_weight = {
```

Comment:
but reading all of this too, perhaps it is easier to make ensure `WeightInfo::unbond` is the worse case, and call it a day, without needing to do any kind of manual weight tracking like this. Will be more or less like this: 

```
#[pallet::weight(T::WeightInfo::unbond_with_withdraw())]
fn unbond(..) {
	..
	..
	..
	if did_withdraw {
		Ok(None)
	} else {
		Ok(unbond_weight_without_withdraw)
	}
}
```

but then you'd have to write a new benchmark.

Both are fine for me for now. I would decide by asking myself which type of code would be more difficult to read for you in 6 months when you will look back at this?

Reply:
Okay, let's keep it as-is then, let's just document `SPECULATIVE_NUM_SPANS` a bit better as noted elsewhere.